### ISSUE TITLE
Feature suggestion: smoothing

### ISSUE THREAD

**Issue by hoangvietanh** [mbloch/mapshaper#336]

**Comment by hoangvietanh** [comment]
A common task in post vector processing is generalization and then smoothing. It would be great if Mapshaper could have option for smoothing.

**Comment by mbloch** [comment]
I agree that smoothing would be a good addition.  Is there a particular smoothing algorithm that you recommend for this?

I've spent a little bit of time thinking about smoothing. Some datasets have a combination of sinuous, curving lines and straight lines. For example, administrative boundaries can follow rivers and other natural features in some areas, and in other areas can be composed of straight lines connected by sharp angles. A good smoothing function should not try to round these sharp angles.

The kind of generalization that mapshaper does (vertex removal) tends to obscure the distinction between curves and sharp angles as the amount of simplification increases. So I'm not sure we'll get the best results if smoothing is applied after generalization, as an independent step.

**Comment by TomBor** [comment]
No idea if it can help but lately I've used this R package called [smoothr](http://strimas.com/smoothr/) to first smooth a vector and then simplify it with mapshaper.

**Comment by mbloch** [comment]
I'll take a look at smoothr, thanks.

**Comment by hoangvietanh** [comment]
Mbloch: Mapshaper is very solid and stable for a long time. Adding smoothing would greatly extend its usability. My smoothing use-case is mostly for raster segmentation/raster classification post processing. So smoothing with this context doesn't need to retain any special feature (such as straight line for admin boundary).

For the algorithm B - Spline could be a good choice because it follows better than the Bezier curve the original shape of the polygon.

For the workflow, it could be good to let user decide to use generalization or smooth as 2 separate task given that each could take input from the other.

Again thank you for making Mapshaper available, it have saved me many hours of work.

**Comment by rowanwins** [comment]
Chaikin's algorithm is super-simple for smoothing polygons
http://graphics.cs.ucdavis.edu/education/CAGDNotes/Chaikins-Algorithm/Chaikins-Algorithm.html

I wrote an implementation for TurfJS ([source code here](https://github.com/Turfjs/turf/blob/master/packages/turf-polygon-smooth/index.js))

Hope that helps :)

**Comment by mbloch** [comment]
Smoothing is now implemented, here's a documentation page about it: https://mapshaper.org/docs/guides/smoothing.html
